src/transformers/models/cogvlm/modeling_cogvlm.py

```diff
@@ -0,0 +1,1155 @@
+        >>> import torch
+        >>> import requests
+        >>> from PIL import Image
+
+        >>> processor = CogvlmProcessor.from_pretrained("THUDM/cogvlm-chat-hf")
+        >>> model = CogvlmModel.from_pretrained("THUDM/cogvlm-chat-hf")
+
+        >>> # load image
+        >>> url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+        >>> image = Image.open(requests.get(url, stream=True).raw)
+        >>> query = "Describe this image"
+
+        >>> prompt = f"Question: {query} Answer:"
+        >>> inputs = processor(images=image, text=prompt, return_tensors="pt")
+
+        >>> # forward pass
+        >>> with torch.no_grad():
+        ...     outputs = model(**inputs)
+
+        >>> last_hidden_state = outputs.last_hidden_state
+        ```
+        """
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        use_cache = use_cache if use_cache is not None else self.config.use_cache
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        if past_key_values is not None:
+            # generate mode with past_key_values. the image features are already mapped
+            # update attention_mask
+            if pixel_values is not None:
+                batch_size = pixel_values.shape[0]
+                vision_mask = torch.full(
+                    size=(batch_size, self.num_vision_tokens), fill_value=1, device=attention_mask.device
+                )
+                attention_mask = torch.cat(
+                    [attention_mask[:, :-1], vision_mask, attention_mask[:, -1].repeat(batch_size, 1)], dim=1
+                )
+        else:
+            # not allow for inputs_embeds, because we want to process image feature
+            # assert input_ids is not None and inputs_embeds is None, f"{input_ids} {inputs_embeds}"
+
+            if pixel_values is not None and input_ids is not None:
+                # multi-modality
+                if len(input_ids) != len(pixel_values):
+                    raise ValueError("Make sure to pass as many texts as images")
+
+                # prepend the input_ids and token_type_ids with image tokens
+                batch_size = input_ids.shape[0]
+
+                vision_input_ids = self.vision_input_ids.repeat(batch_size, 1)
+                vision_input_ids = vision_input_ids.to(input_ids.device)
+
+                vision_token_type_ids = self.vision_token_type_ids.repeat(batch_size, 1)
+                vision_token_type_ids = vision_token_type_ids.to(token_type_ids.device)
+
+                input_ids = torch.cat([vision_input_ids, input_ids[:, 1:]], dim=1)
+                token_type_ids = torch.cat([vision_token_type_ids, token_type_ids[:, 1:]], dim=1)
```

Comment:
Yes currently token_type_ids are created inside the model.

I assume that if we want to take the same route as Idefics2, that the processor should create these. Currently the processor just consists of `CLIPImageProcessor` and `LlamaTokenizer`, and the processor only creates `pixel_values`, `input_ids` and `attention_mask`. The `token_type_ids` are created inside the model.

Reply:
I realized that's how I designed it in the beginning 😭 then I moved the `token_type_ids` calculation inside the model so that it would work with the `image-to-text` pipeline (which only relies on the image processor and tokenizer separately).

However I assume we can now move that logic back again to the processor class.

Reply:
The first implementation I had is here: https://github.com/NielsRogge/transformers/blob/add_cogvlm/src/transformers/models/cogvlm/processing_cogvlm.py#L86-L133.

Let me know what you prefer :)

Reply:
Thanks, now that I see the old version I got the model logic 😄 

For token type ids, my vote will go for preparing all inputs in the "preprocessor" for consistency. Also the concern is about `token_type_ids[:, 1:]` part that should give error if `token_type_ids` are not passed in multimodal `forward`

I will leave for you and @amyeroberts to decide what is best

Reply:
Preparing the token_type_ids in the processor has my vote!

Reply:
Now we prepare all inputs in the model.